From ed79c3432d85d11c6d59933fdc441294d9df6f4a Mon Sep 17 00:00:00 2001
From: Andrej Shadura <andrew.shadura@collabora.co.uk>
Date: Thu, 15 Nov 2018 10:12:46 -0800
Subject: [PATCH] When --sysroot is being used, we want absolute paths to refer
 to directories inside the sysroot.

There's subtle difference in parsing of /etc/ld.so.conf by ldconfig and ld:

* ldconfig interprets relative paths in includes relatively to the
  current directory it run in, and absolute path relatively to the
  current "chroot" directory, if set

* ld interprets relative paths in includes relatively to the directory
  the parsed file is in, and apparently ignores the --sysroot setting
  for absolute paths and does not relativise them to this sysroot.

Using relative paths in ld.so.conf breaks ldconfig, using absolute paths there
breaks ld. This patch changes the behaviour of ld in this situation to match
that of ldconfig.

Signed-off-by: Andrej Shadura <andrew.shadura@collabora.co.uk>
---
 ld/emultempl/elf32.em | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/ld/emultempl/elf32.em b/ld/emultempl/elf32.em
index ad31a625ef..63d026f734 100644
--- a/ld/ldelf.c
+++ b/ld/ldelf.c
@@ -795,6 +795,10 @@ gld${EMULATION_NAME}_parse_ld_so_conf_include
       memcpy (newp + (p - filename + 1), pattern, patlen);
       pattern = newp;
     }
+  else
+    {
+      pattern = concat (ld_sysroot, pattern, NULL);
+    }
 
 #ifdef HAVE_GLOB
   if (glob (pattern, 0, NULL, &gl) == 0)
-- 
2.19.1

